pkg/agent/openflow/pipeline.go

```diff
@@ -454,388 +502,471 @@ func (c *client) DeleteOFEntries(ofEntries []binding.OFEntry) error {
+// Tables: SNATConntrackTable
+// Stage: stagePostRouting
+// Tables: SNATConntrackCommitTable
+// Refactored from:
+//   - part of func (c *client) connectionTrackFlows(category cookie.Category) []binding.Flow
+// snatConntrackFlows generates the flows about conntrack of SNAT connection for feature Service .
+func (f *featureService) snatConntrackFlows() []binding.Flow {
+	cookieID := f.cookieAllocator.Request(f.category).Raw()
 	var flows []binding.Flow
-	if packet == nil {
-		for _, ipProtocol := range c.ipProtocols {
-			flowBuilder := ConntrackStateTable.BuildFlow(priorityLow + 1).
+	for _, ipProtocol := range f.ipProtocols {
+		flows = append(flows,
+			// This generates the flow to transform the packets from tracked SNATed Service connection in SNAT CT zone.
+			SNATConntrackTable.ofTable.BuildFlow(priorityNormal).
+				Cookie(cookieID).
 				MatchProtocol(ipProtocol).
-				MatchIPDSCP(dataplaneTag).
-				SetHardTimeout(timeout).
-				Cookie(c.cookieAllocator.Request(category).Raw())
-			if c.enableProxy {
-				flowBuilder = flowBuilder.
-					Action().ResubmitToTable(SessionAffinityTable.GetID()).
-					Action().ResubmitToTable(ServiceLBTable.GetID())
-			} else {
-				flowBuilder = flowBuilder.
-					Action().ResubmitToTable(ConntrackStateTable.GetNext())
-			}
-			flows = append(flows, flowBuilder.Done())
+				Action().CT(false, SNATConntrackTable.ofTable.GetNext(), f.snatCtZones[ipProtocol]).
+				NAT().
+				CTDone().
+				Done(),
 
-			flows = append(flows, ConntrackStateTable.BuildFlow(priorityLow+2).
+			// SNAT should be performed for the following connections:
+			// - Hairpin Service connection initiated through a local Pod, and SNAT should be performed with the Antrea
+			//   gateway IP.
+			// - Hairpin Service connection initiated through the Antrea gateway, and SNAT should be performed with a
+			//   virtual IP.
+			// - Nodeport / LoadBalancer connection initiated through the Antrea gateway and externalTrafficPolicy is
+			//   Cluster, and SNAT should be performed with the Antrea gateway IP.
+
+			// This generates the flow to match the first packet of hairpin Service connection initiated through the Antrea
+			// gateway with ConnSNATCTMark and HairpinCTMark, then perform SNAT in SNAT CT zone with a virtual IP.
+			SNATConntrackCommitTable.ofTable.BuildFlow(priorityNormal).
+				Cookie(cookieID).
 				MatchProtocol(ipProtocol).
-				MatchIPDSCP(dataplaneTag).
-				MatchCTStateTrk(true).MatchCTStateRpl(true).
-				SetHardTimeout(timeout).
-				Cookie(c.cookieAllocator.Request(category).Raw()).
-				Action().Drop().
-				Done())
+				MatchCTStateNew(true).
+				MatchCTStateTrk(true).
+				MatchRegMark(FromGatewayRegMark).
+				MatchCTMark(ConnSNATCTMark).
+				MatchCTMark(HairpinCTMark).
+				Action().CT(true, SNATConntrackCommitTable.ofTable.GetNext(), f.snatCtZones[ipProtocol]).
+				SNAT(&binding.IPRange{StartIP: f.virtualIPs[ipProtocol], EndIP: f.virtualIPs[ipProtocol]}, nil).
+				LoadToCtMark(ServiceCTMark).
```

Comment:
I saw that `ServiceCTMark` is matched in table `ConntrackStateTable`, `L3ForwardingTable`, and `ConntrackCommitTable`. I think the point to user snat zone mark should be in `ConntrackCommitTable` to avoid commit Service traffic in conntrackZone duplicatedly. And using this special mark instead of `ct_state=+dnat` is because the dnat state is lost in the snat zone, right? If yes, could you adding comment here to explain why we must copy this mark in snat zone from dnat zone?

Reply:
Added.